controllers/operatorpolicy_controller.go

```diff
@@ -60,166 +65,205 @@ var _ reconcile.Reconciler = &OperatorPolicyReconciler{}
 			}
 
-			r.setCompliance(policy, policyv1.Compliant)
+			return reconcile.Result{Requeue: true}, err
+		} else if policy.Spec.OperatorGroup != nil {
+			// If inform mode, keep going and return at the end without requeue. Before
+			// continuing, should check if required og spec is created to set compliance state
 
-			return nil
+			// Set to non compliant because operatorgroup does not exist on cluster, but
+			// should still go to check subscription
+			r.setCompliance(policy, policyv1.NonCompliant)
 		}
+	}
 
-		// Inform
-		r.setCompliance(policy, policyv1.NonCompliant)
-
-		return nil
+	// Retrieve subscription from cache
+	cachedSubscription, err := r.DynamicWatcher.GetFromCache(
+		schema.GroupVersionKind{Group: "operators.coreos.com", Version: "v1alpha1", Kind: "Subscription"},
+		policy.Spec.Subscription.Namespace, policy.Spec.Subscription.SubscriptionSpec.Package)
+	if err != nil {
+		panic(err)
 	}
 
-	// Object exists but it should not exist, delete object
-	// Deleting related objects will be added in the future
-	if exists && !shouldExist {
+	// Create new subscription
+	if cachedSubscription == nil {
 		if strings.EqualFold(string(policy.Spec.RemediationAction), string(policyv1.Enforce)) {
-			OpLog.Info("deleting kind " + subscriptionSpec.Kind + " in ns " + subscriptionSpec.Namespace)
-			err = r.Delete(context.TODO(), subscriptionSpec)
+			subscriptionSpec := buildSubscription(policy)
+			OpLog.Info("creating kind " + subscriptionSpec.Kind + " in ns " + subscriptionSpec.Namespace)
 
+			err := r.Create(context.TODO(), subscriptionSpec)
 			if err != nil {
 				r.setCompliance(policy, policyv1.NonCompliant)
-				OpLog.Error(err, "Could not handle existing musthave object")
+				OpLog.Error(err, "Could not handle missing musthave object")
+
+				err = r.DynamicWatcher.EndQueryBatch(watcher)
+				if err != nil {
+					panic(err)
+				}
 
-				return err
+				return reconcile.Result{}, err
 			}
 
 			r.setCompliance(policy, policyv1.Compliant)
 
-			return nil
-		}
+			err = r.DynamicWatcher.EndQueryBatch(watcher)
+			if err != nil {
+				panic(err)
+			}
 
-		// Inform
+			return reconcile.Result{Requeue: true}, nil
+		}
+		// inform
 		r.setCompliance(policy, policyv1.NonCompliant)
-
-		return nil
 	}
 
-	// Object does not exist and it should not exist, emit success event
-	if !exists && !shouldExist {
-		OpLog.Info("The object does not exist and is compliant with the mustnothave compliance type")
-		// Future implementation: Possibly emit a success event
-
-		return nil
-	}
-
-	// Object exists, now need to validate field to make sure they match
-	if exists {
-		OpLog.Info("The object already exists. Checking fields to verify matching specs")
-		// Future implementation: Verify the specs of the object matches the one on the cluster
-
-		return nil
+	err = r.DynamicWatcher.EndQueryBatch(watcher)
+	if err != nil {
+		panic(err)
 	}
 
-	return nil
+	// Will only reach this if in inform mode
+	return reconcile.Result{}, nil
```

Comment:
It seems like there's an 'extra' reconcile that occurs after this line is called in inform mode. Not sure if I'm missing some details but would appreciate some insight on why this happens.

Reply:
This also happens in enforce mode, but I'm assuming that's because the subscription object changes during the process of installing the operator.

Reply:
Two thoughts, one or both might be incorrect though 😆 

1. The controller watches for changes to the OperatorPolicy. If a reconcile loop causes a status change on the OperatorPolicy, that might trigger another reconcile immediately. In some controllers, we deliberately set things up so that status-only changes do not cause a new reconcile, but I don't think we've done that here.

2. The "first" reconcile is caused by the creation of the OperatorPolicy. During that reconcile, regardless of inform/enforce or anything else it does, it creates watches on the OperatorGroups and Subscriptions (ideally just the ones concerning this policy, but I think the current implementation is not that specific). When those watches are created, they sent an initial event that may cause `Reconcile` to be called.

Reply:
I think the first thought is correct because they are setting `DisableInitialReconcile: true` on the dependency watcher.

You should be able to use a predicate to ignore status updates. I think controller-runtime has a built-in one for just checking if the "generation" changes (i.e. `spec` has changed).